src/NumericalAlgorithms/Interpolation/MultiLinearSpanInterpolation.hpp

```diff
@@ -0,0 +1,397 @@
+size_t MultiLinearSpanInterpolation<
+    Dimension, NumberOfVariables,
+    UniformSpacing>::findIndexGeneral(const size_t whichDimension,
+                                      const double& targetPoints) const {
+  size_t lowerIndexBound = 0;
+  size_t upperIndexBound = numberOfPoints[whichDimension] - 1;
+
+  ASSERT((targetPoints > x[whichDimension][lowerIndexBound]) or
+             allow_extrapolation_below_data[whichDimension],
+         "Interpolation exceeds lower table bounds");
+  ASSERT((targetPoints < x[whichDimension][upperIndexBound]) or
+             allow_extrapolation_above_data[whichDimension],
+         "Interpolation exceeds upper table bounds");
+
+  while (upperIndexBound > 1 + lowerIndexBound) {
+    size_t currentIndex =
+        lowerIndexBound + (upperIndexBound - lowerIndexBound) / 2;
+    if (targetPoints < x[whichDimension][currentIndex])
+      upperIndexBound = currentIndex;
+    else
+      lowerIndexBound = currentIndex;
+  }
+  return lowerIndexBound;
+
+  // if (targetPoints < x[whichDimension][0])
+  //   return lowerIndexBound;
+  // if (targetPoints > x[whichDimension][upperIndexBound])
+  //   return upperIndexBound;
+  //
+  // return std::distance(x[whichDimension].begin(),
+  //                      std::lower_bound(x[whichDimension].begin(),
+  //                                       x[whichDimension].end(),
+  //                                       targetPoints));
+}
+
+//! Determine relative index bracketing for non-uniform spacing
+template <size_t Dimension, size_t NumberOfVariables, bool UniformSpacing>
+size_t MultiLinearSpanInterpolation<
+    Dimension, NumberOfVariables,
+    UniformSpacing>::findIndexUniform(const size_t whichDimension,
+                                      const double& targetPoints) const {
+  // Compute coordinate relative to lowest table bound
+  const auto relativeCoordinate = (targetPoints - x[whichDimension][0]);
+
+  auto currentIndex =
+      static_cast<size_t>(relativeCoordinate * inverseSpacing[whichDimension]);
+
+  // We are exceeding the table bounds:
+  // Use linear extrapolation based of the lowest
+  // two points in the table
+  ASSERT(allow_extrapolation_below_data[whichDimension] or
+             UNLIKELY(relativeCoordinate > 0.),
+         "Interpolation exceeds lower table bounds.");
+
+  // We are exceeding the table bounds:
+  // Use linear extrapolation based of the highest
+  // two points in the table
+
+  ASSERT(allow_extrapolation_above_data[whichDimension] or
+             UNLIKELY(currentIndex + 1 < numberOfPoints[whichDimension]),
+         "Interpolation exceeds upper table bounds.");
+
+  // Enforce index ranges
+  currentIndex =
+      std::min(numberOfPoints[whichDimension] - 2, std::max(0ul, currentIndex));
+
+  return currentIndex;
+}
+
+/// Compute interpolation weights for 1D tables
+template <size_t Dimension, size_t NumberOfVariables, bool UniformSpacing>
+typename MultiLinearSpanInterpolation<Dimension, NumberOfVariables,
```

Comment:
I think if you use trailing return type syntax you can just do `Weight<1>`. I.e. `auto Mult...::get_weights() const -> Weight<1>` (I always forget if the `const` goes before or after the return type, sorry...)